### PR TITLE
Feature/47 local storage

### PR DESCRIPTION
**Pull request by alexander-toch**
## To Do

- [x] save selected language
- [x] save selected newsgroups in DB
- [x] integrate DB Build into Splash screen (don't show the Add Server Fragment before navigating to the List)
- [x] delete Server functionality

### CODE REVIEW

File: app/src/androidTest/java/com/tugraz/asd/modernnewsgroupapp/EditNewsgroupAliasTest.kt [hunk 1]
```diff
@@ -47,6 +49,11 @@ class EditNewsgroupAliasTest {
...
     }
 
+    @Before
+    fun clearDb(){
+        InstrumentationRegistry.getInstrumentation().getTargetContext().deleteDatabase("newsgroup.db")
+    }
+
     @Test
     fun checkIfNewsgroupServerDisplayed()
     {
```
Comment (PatrickStru) [inline app/src/androidTest/java/com/tugraz/asd/modernnewsgroupapp/EditNewsgroupAliasTest.kt:59]: Testcase not working.

File: app/src/androidTest/java/com/tugraz/asd/modernnewsgroupapp/DeleteLastServerTest.kt [hunk 2]
```diff
@@ -79,7 +86,7 @@ class DeleteLastServerTest {
 
         val checkBox = onView(
             allOf(
-                withText("tu-graz.algorithmen"),
+                withText("vc-graz"),
                 childAtPosition(
                     allOf(
                         withId(R.id.view_subscribe),
```
Comment (PatrickStru) [inline app/src/androidTest/java/com/tugraz/asd/modernnewsgroupapp/DeleteLastServerTest.kt:92]: Testcase not working

File: app/src/androidTest/java/com/tugraz/asd/modernnewsgroupapp/NewsgroupServerMenuTest.kt [hunk 3]
```diff
@@ -25,7 +27,12 @@ class NewsgroupServerMenuTest {
...
-    var mActivityTestRule = ActivityTestRule(SplashScreen::class.java)
+    var mActivityTestRule = ActivityTestRule(MainActivity::class.java)
+
+    @Before
+    fun clearDb(){
+        InstrumentationRegistry.getInstrumentation().getTargetContext().deleteDatabase("newsgroup.db")
+    }
 
     @Test
     fun newsgroupServerMenuTest() {
```
Comment (PatrickStru) [inline app/src/androidTest/java/com/tugraz/asd/modernnewsgroupapp/NewsgroupServerMenuTest.kt:38]: Testcase not working

File: app/src/androidTest/java/com/tugraz/asd/modernnewsgroupapp/ServerMenuTest.kt [hunk 4]
```diff
@@ -25,7 +27,12 @@ class ServerMenuTest {
...
-    var mActivityTestRule = ActivityTestRule(SplashScreen::class.java)
+    var mActivityTestRule = ActivityTestRule(MainActivity::class.java)
+
+    @Before
+    fun clearDb(){
+        InstrumentationRegistry.getInstrumentation().getTargetContext().deleteDatabase("newsgroup.db")
+    }
 
     @Test
     fun serverMenuTest() {
```
Comment (PatrickStru) [inline app/src/androidTest/java/com/tugraz/asd/modernnewsgroupapp/ServerMenuTest.kt:38]: Testcase not working

File: app/src/androidTest/java/com/tugraz/asd/modernnewsgroupapp/ShowSubscribedNewsgroupTest.kt [hunk 5]
```diff
@@ -29,6 +31,11 @@ class ShowSubscribedNewsgroupTest {
...
     }
 
+    @Before
+    fun clearDb(){
+        InstrumentationRegistry.getInstrumentation().getTargetContext().deleteDatabase("newsgroup.db")
+    }
+
     @Test
     fun showSubscribedNewsgroupTest() {
         init()
```
Comment (PatrickStru) [inline app/src/androidTest/java/com/tugraz/asd/modernnewsgroupapp/ShowSubscribedNewsgroupTest.kt:41]: SQLite Exception

File: app/src/androidTest/java/com/tugraz/asd/modernnewsgroupapp/SnackBarTests.kt [hunk 6]
```diff
@@ -17,7 +19,12 @@ class SnackBarTests {
...
-    var  rule: ActivityScenarioRule<ActivityAddNewsgroup> = ActivityScenarioRule<ActivityAddNewsgroup>(ActivityAddNewsgroup::class.java)
+    var  rule: ActivityScenarioRule<MainActivity> = ActivityScenarioRule<MainActivity>(MainActivity::class.java)
+
+    @Before
+    fun clearDb(){
+        InstrumentationRegistry.getInstrumentation().getTargetContext().deleteDatabase("newsgroup.db")
+    }
 
     @Test
     fun checkIfSnackBarIsDisplayed()
```
Comment (PatrickStru) [inline app/src/androidTest/java/com/tugraz/asd/modernnewsgroupapp/SnackBarTests.kt:30]: SQLite Exception

File: app/src/androidTest/java/com/tugraz/asd/modernnewsgroupapp/SubgroupInstrumentedTest.kt [hunk 7]
```diff
@@ -32,6 +34,11 @@ class SubgroupInstrumentedTest {
...
     }
 
+    @Before
+    fun clearDb(){
+        InstrumentationRegistry.getInstrumentation().getTargetContext().deleteDatabase("newsgroup.db")
+    }
+
     @Test
     fun checkIfListCollapses() {
         init()
```
Comment (PatrickStru) [inline app/src/androidTest/java/com/tugraz/asd/modernnewsgroupapp/SubgroupInstrumentedTest.kt:44]: SQLite Exception

File: app/src/androidTest/java/com/tugraz/asd/modernnewsgroupapp/SwipeSubgroupListTest.kt [hunk 8]
```diff
@@ -21,7 +21,7 @@ class SwipeSubgroupListTest {
 
     @Rule
     @JvmField
-    var  rule: ActivityScenarioRule<ActivityAddNewsgroup> = ActivityScenarioRule(ActivityAddNewsgroup::class.java)
+    var  rule: ActivityScenarioRule<MainActivity> = ActivityScenarioRule(MainActivity::class.java)
 
     private fun init() {
         val inputName = onView(ViewMatchers.withId(R.id.editText_name)).check(matches(isDisplayed()))
```
Comment (PatrickStru) [inline app/src/androidTest/java/com/tugraz/asd/modernnewsgroupapp/SwipeSubgroupListTest.kt:27]: Testcase not working

File: app/src/main/java/com/tugraz/asd/modernnewsgroupapp/FragmentProfile.kt [hunk 9]
```diff
@@ -80,7 +81,16 @@ class FragmentProfile : Fragment() {
...
+                if (prefs != null) {
+                    with (prefs.edit()) {
+                        putString(getString(R.string.language_pref_key), lang)
+                        apply()
+                    }
+                }
+
                 startActivity(refresh)
             } else {
                 findNavController().navigate(R.id.action_FragmentProfile_to_FragmentShowSubgroups)
```
Comment (PatrickStru) [inline app/src/main/java/com/tugraz/asd/modernnewsgroupapp/FragmentProfile.kt:96]: Language switch is applied after second attempt

File: app/src/main/java/com/tugraz/asd/modernnewsgroupapp/FragmentEditNewsgroup.kt [hunk 10]
```diff
@@ -55,16 +50,30 @@ class FragmentEditNewsgroup : Fragment() {
...
+                    findNavController().navigate(R.id.action_FragmentEditNewsgroup_to_FragmentShowSubgroups)
+                }
+            } else {
+                withContext(Dispatchers.Main){
+                    findNavController().navigate(R.id.action_FragmentEditNewsgroup_to_FragmentAddNewsgroup)
+                }
+            }
+        }
         Feedback.showSuccess(this.requireView(), "Newsgroup Server successfully deleted.")
     }
```
Comment (PatrickStru) [inline app/src/main/java/com/tugraz/asd/modernnewsgroupapp/FragmentEditNewsgroup.kt:79]: Newsgroup Server Alias not updated after renaming

File: app/src/main/java/com/tugraz/asd/modernnewsgroupapp/MainActivity.kt [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    override fun attachBaseContext(newBase: Context) {
+        // get chosen language from shared preference
+        val prefs = PreferenceManager.getDefaultSharedPreferences(newBase);
+        val localeString = prefs.getString("language", "en")
+        println("Language: $localeString")
+        val localeUpdatedContext: ContextWrapper = ContextUtils.updateLocale(newBase, Locale(localeString))
+        super.attachBaseContext(localeUpdatedContext)
+    }
+
+}
```
Comment (PatrickStru) [inline app/src/main/java/com/tugraz/asd/modernnewsgroupapp/MainActivity.kt:81]: If I already added a newsgroup server before there is no way to add another newsgroup server anymore due to skipSetup?

**Review by PatrickStru — COMMENTED** [review]
Besides not properly working testcases the language switch and skipping the setup should be rethought.

File: app/src/main/java/com/tugraz/asd/modernnewsgroupapp/MainActivity.kt (hunk 11 above)
Comment (alexander-toch) [reply app/src/main/java/com/tugraz/asd/modernnewsgroupapp/MainActivity.kt:81]: there also was no way before, besides going back with the back button, which is also not the correct behavior.
Not in the scope of this ticket.

File: app/src/main/java/com/tugraz/asd/modernnewsgroupapp/FragmentEditNewsgroup.kt (hunk 10 above)
Comment (alexander-toch) [reply app/src/main/java/com/tugraz/asd/modernnewsgroupapp/FragmentEditNewsgroup.kt:79]: not in the scope of this ticket, but easy to implement tbh

**Comment by alexander-toch** [comment]
> Besides not properly working testcases the language switch and skipping the setup should be rethought.

I think most of the testcases also were not working before anymore (just check the current master). 
How should we proceed here @jakobdunst? As I don;t have the additional 4 hours anymore, maybe someone of the master students in our group could take this over?

**Review by PatrickStru — APPROVED** [review]

**Pull request merged**